pkg/agent/flowexporter/exporter/l7Listener.go

```diff
@@ -0,0 +1,189 @@
+// Copyright 2023 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporter
+
+import (
+	"bufio"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"k8s.io/klog/v2"
+
+	"antrea.io/antrea/pkg/agent/flowexporter"
+)
+
+type l7Listener struct {
+	l7EventsHttp map[flowexporter.ConnectionKey]map[int32]Http
+	l7mut        sync.Mutex
+}
+
+func NewL7Listener() *l7Listener {
+	return &l7Listener{
+		l7EventsHttp: make(map[flowexporter.ConnectionKey]map[int32]Http),
+	}
+}
+
+func (l *l7Listener) Run(stopCh <-chan struct{}, socketPath string) {
+	listener, err := func() (net.Listener, error) {
+		if err := os.Remove(socketPath); err != nil && !os.IsNotExist(err) {
+			return nil, fmt.Errorf("failed to remove stale socket: %w", err)
+		}
+		if err := os.MkdirAll(filepath.Dir(socketPath), 0755); err != nil {
+			return nil, fmt.Errorf("failed to create directory %s: %w", socketPath, err)
+		}
+		listener, err := net.Listen("unix", socketPath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to listen on socket: %w", err)
+		}
+		return listener, nil
+	}()
+	if err != nil {
+		klog.Fatalf("Failed to initialize L7Listener: %v", err)
+	}
+	defer listener.Close()
+
+	klog.InfoS("L7 Listener Server started. Listening for connections...")
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			conn, err := listener.Accept()
+			if err != nil {
+				klog.ErrorS(err, "Error accepting Suricata connection")
+				continue
+			}
+			go l.handleClientConnection(conn)
+		}
+	}
+}
+
+func (l *l7Listener) handleClientConnection(conn net.Conn) {
+	defer conn.Close()
+	reader := bufio.NewReader(conn)
+	for {
+		buffer, err := reader.ReadBytes('\n')
+		if err != nil && err != io.EOF {
+			klog.ErrorS(err, "Error reading data", "buffer", buffer)
+			return
+		}
+		err = l.processLog(buffer)
+		if err != nil {
+			klog.ErrorS(err, "Error while processing L7 data")
+			return
+		}
+	}
+}
+
+func (l *l7Listener) processLog(data []byte) error {
+	l.l7mut.Lock()
+	defer l.l7mut.Unlock()
+	var eventType string
+	// Check if the event type is "http"
+	if eventType = extractEventType(data); eventType != "http" {
+		return nil
+	}
+	var event JsonToEvent
+	err := json.Unmarshal(data, &event)
```

Comment:
I don't get why it unmarshals the string twice. It will just duplicate the overhead when the event is http. `extractEventType` shouldn't be there at all, you can do a benchmark test but I think defining a temporary struct as receiver doesn't mean `json.Unmarshal` doesn't parse the whole string.

Reply:
yes you are correct!, removing it!